pennylane/estimator/resource_operator.py

```diff
@@ -0,0 +1,442 @@
+
+    @classmethod
+    def adjoint_resource_decomp(cls, *args, **kwargs) -> list:
+        r"""Returns a list representing the resources for the adjoint of the operator."""
+        raise ResourcesNotDefined
+
+    @classmethod
+    def controlled_resource_decomp(
+        cls, ctrl_num_ctrl_wires: int, ctrl_num_ctrl_values: int, *args, **kwargs
+    ) -> list:
+        r"""Returns a list representing the resources for a controlled version of the operator.
+
+        Args:
+            ctrl_num_ctrl_wires (int): the number of qubits the
+                operation is controlled on
+            ctrl_num_ctrl_values (int): the number of control qubits, that are
+                controlled when in the :math:`|0\rangle` state
+        """
+        raise ResourcesNotDefined
+
+    @classmethod
+    def pow_resource_decomp(cls, pow_z: int, *args, **kwargs) -> list:
+        r"""Returns a list representing the resources for an operator
+        raised to a power.
+
+        Args:
+            pow_z (int): exponent that the operator is being raised to
+        """
+        raise ResourcesNotDefined
+
+    def __repr__(self) -> str:
+        str_rep = self.__class__.__name__ + "(" + str(self.resource_params) + ")"
+        return str_rep
+
+    def __mul__(self, scalar: int):
+        assert isinstance(scalar, int)
+        gate_types = defaultdict(int, {self.resource_rep_from_op(): scalar})
+        wire_manager = WireResourceManager(0, algo=self.num_wires)
+
+        return Resources(wire_manager, gate_types)
+
+    def __matmul__(self, scalar: int):
+        assert isinstance(scalar, int)
+        gate_types = defaultdict(int, {self.resource_rep_from_op(): scalar})
+        wire_manager = WireResourceManager(0, algo=scalar * self.num_wires)
+
+        return Resources(wire_manager, gate_types)
+
+    def __add__(self, other):
+        if isinstance(other, ResourceOperator):
+            return (1 * self) + (1 * other)
+        if isinstance(other, Resources):
+            return (1 * self) + other
+
+        raise TypeError(f"Cannot add resource operator {self} with type {type(other)}.")
+
+    def __and__(self, other):
+        if isinstance(other, ResourceOperator):
+            return (1 * self) & (1 * other)
+        if isinstance(other, Resources):
+            return (1 * self) & other
+
+        raise TypeError(f"Cannot add resource operator {self} with type {type(other)}.")
+
+    @classmethod
+    def tracking_name(cls, *args, **kwargs) -> str:
+        r"""Returns a name used to track the operator during resource estimation."""
+        return cls.__name__.replace("Resource", "")
+
+    def tracking_name_from_op(self) -> str:
+        r"""Returns the tracking name built with the operator's parameters."""
+        return self.__class__.tracking_name(**self.resource_params)
+
+
+class ResourcesNotDefined(Exception):
+    r"""Exception to be raised when a ``ResourceOperator`` does not implement _resource_decomp"""
```

Comment:
Can we place this in the `exceptions` module? Helps us keep track of all the exceptions we've defined in all the different places.